polynote-server/src/main/scala/polynote/server/SharedNotebook.scala

```diff
@@ -0,0 +1,285 @@
+}
+
+
+class IOSharedNotebook(
+  val path: String,
+  ref: SignallingRef[IO, (GlobalVersion, Notebook)],            // the Int is the global version, which can wrap around back to zero if necessary
+  kernelRef: Ref[IO, Option[Kernel[IO]]],
+  updates: Queue[IO, Option[(SubscriberId, NotebookUpdate, Deferred[IO, GlobalVersion])]],   // the canonical series of edits
+  updatesTopic: Topic[IO, Option[(GlobalVersion, SubscriberId, NotebookUpdate)]],  // a subscribe-able channel for watching updates,
+  kernelFactory: KernelFactory[IO],
+  outputMessages: Topic[IO, Message],
+  kernelLock: Semaphore[IO]
+)(implicit contextShift: ContextShift[IO]) extends SharedNotebook[IO] {
+
+  private val updateBuffer = new VersionBuffer[NotebookUpdate]
+
+  private val statusUpdates = Publish.PublishTopic(outputMessages).contramap[KernelStatusUpdate](KernelStatus(ShortString(path), _))
+
+  // listen to the stream of updates and apply them in order, each one incrementing the global version
+  updates.dequeue.unNoneTerminate.zipWithIndex.evalMap {
+    case ((subscriberId, update, versionPromise), version) =>
+      val newGlobalVersion = (version % Int.MaxValue).toInt
+      updateBuffer.add(newGlobalVersion, update)
+      applyUpdate(newGlobalVersion, subscriberId, update, versionPromise)
+  }.compile.drain.unsafeRunAsyncAndForget()
+
+  private def ensureKernel(): IO[Kernel[IO]] = kernelLock.acquire.bracket { _ =>
+    kernelRef.get.flatMap {
+      case None => kernelFactory.launchKernel(() => ref.get.map(_._2), statusUpdates).flatMap {
+        kernel => kernelRef.set(Some(kernel)).as(kernel)
+      }
+      case Some(kernel) => IO.pure(kernel)
+    }
+  }(_ => kernelLock.release)
+
+  // apply a versioned update from the queue, completing its version promise and updating the info about which
+  // versions the originating subscriber knows about
+  private def applyUpdate(newGlobalVersion: Int, subscriberId: Int, update: NotebookUpdate, versionPromise: Deferred[IO, GlobalVersion]) = ref.get.flatMap {
+    case (currentVer, notebook) =>
+
+      // TODO: remove this, just checking for now
+      assert(newGlobalVersion - 1 == currentVer, "Version is wrong!")
+
+      val doUpdate = update match {
+        case InsertCell(_, _, _, cell, after) => ref.set(newGlobalVersion -> notebook.insertCell(cell, after))
+        case DeleteCell(_, _, _, id)          => ref.set(newGlobalVersion -> notebook.deleteCell(id))
+        case UpdateCell(_, _, _, id, edits)   => ref.set(newGlobalVersion -> notebook.editCell(id, edits))
+        case UpdateConfig(_, _, _, config)    => ref.set(newGlobalVersion -> notebook.copy(config = Some(config)))
+        case SetCellLanguage(_, _, _, id, lang) => ref.set(newGlobalVersion -> notebook.updateCell(id)(_.copy(language = lang)))
+      }
+
+      for {
+        _  <- doUpdate
+        sub = subscribers.get(subscriberId)
+        _  <- if (sub != null) sub.setKnownVersions(update.globalVersion, update.localVersion) else IO.unit
```

Comment:
Hmm, if we can't find the subscriber wouldn't that be an error? When would that happen?

Reply:
I was thinking that eventually the subscriber might go away (i.e. when they close their websocket), and that wouldn't really be an error. If the subscriber is gone we don't have to keep track of their versions anymore, but we could still be processing an update from them.

Reply:
I see, that makes sense